packages/dds/tree/src/test/shared-tree/sharedTreeFuzzTests.ts

```diff
@@ -0,0 +1,361 @@
+}
+
+export async function performFuzzActionsAbort(
+    generator: AsyncGenerator<Operation, FuzzTestState>,
+    seed: number,
+    validateAnchor?: boolean,
+    saveInfo?: { saveAt?: number; saveOnFailure: boolean; filepath: string },
+): Promise<Required<FuzzTestState>> {
+    const random = makeRandom(seed);
+    const provider = await TestTreeProvider.create(1, SummarizeType.onDemand);
+
+    const initialTreeState: JsonableTree = {
+        type: brand("Node"),
+        fields: {
+            foo: [
+                { type: brand("Number"), value: 0 },
+                { type: brand("Number"), value: 1 },
+                { type: brand("Number"), value: 2 },
+            ],
+            foo2: [
+                { type: brand("Number"), value: 0 },
+                { type: brand("Number"), value: 1 },
+                { type: brand("Number"), value: 2 },
+            ],
+        },
+    };
+
+    const tree = provider.trees[0];
+
+    initializeTestTree(provider.trees[0], initialTreeState);
+    validateTree(provider.trees[0], [initialTreeState]);
+
+    // building the anchor for anchor stability test
+    const cursor = tree.forest.allocateCursor();
+    moveToDetachedField(tree.forest, cursor);
+    cursor.enterNode(0);
+    cursor.getPath();
+    cursor.firstField();
+    cursor.getFieldKey();
+    cursor.enterNode(1);
+    const firstAnchor = cursor.buildAnchor();
+    cursor.free();
+
+    const initialState: FuzzTestState = {
+        random,
+        testTreeProvider: provider,
+        numberOfEdits: 0,
+    };
+    const finalState = await performFuzzActionsBase(
+        generator,
+        {
+            edit: async (state, operation) => {
+                const { index, contents } = operation;
+                abortFuzzChange(tree, contents);
+                return state;
+            },
+            synchronize: async (state) => {
+                const { testTreeProvider } = state;
+                if (testTreeProvider === undefined) {
+                    fail("Attempted to synchronize with undefined testObjectProvider");
+                }
+                await testTreeProvider.ensureSynchronized();
+                return state;
+            },
+        },
+        initialState,
+        saveInfo,
+    );
+
+    await finalState.testTreeProvider.ensureSynchronized();
+    validateTree(provider.trees[0], [initialTreeState]);
+
+    if (validateAnchor) {
+        const expectedPath: UpPath = {
+            parent: {
+                parent: undefined,
+                parentIndex: 0,
+                parentField: rootFieldKeySymbol,
+            },
+            parentField: brand("foo"),
+            parentIndex: 1,
+        };
+        const anchorPath = tree.locate(firstAnchor);
+        assert(compareUpPaths(expectedPath, anchorPath));
+    }
+    return finalState as Required<FuzzTestState>;
+}
+
+function validateTree(tree: ISharedTree, expected: JsonableTree[]): void {
```

Comment:
Pull this into a utility file so both this file and sharedTree.spec.ts can use it.